src/runtime/interrupt/interrupt_esp32c3.go

```diff
@@ -0,0 +1,198 @@
+// +build esp32c3
+
+package interrupt
+
+import (
+	"device/esp"
+	"device/riscv"
+	"errors"
+	"runtime/volatile"
+	"unsafe"
+)
+
+const (
+	// Each module of the ESP32C3 uses different interrupt registers.
+	// Below is the list of modules used from machine package to initialize interrupt handler
+	GPIO HandlerID = iota
+	UART0
+	UART1
+	USB
+	I2C0
+	I2S1
+	TIMG0
+	TIMG1
+	UHCI0
+	RMT
+	SPI1
+	SPI2
+	TWAI
+	RNG
+	WIFI
+	BT
+	WIFI_BT_COMMON
+	BT_BASEBAND
+	BT_LC
+	RSA
+	AES
+	SHA
+	HMAC
+	DS
+	GDMA
+	SYSTIMER
+	SARADC
+)
+
+var (
+	ErrInterruptNotImplemented = errors.New("interrupt number is not implemented")
+)
+
+type HandlerID int
+
+func (i Interrupt) Enable() error {
+	mask := riscv.DisableInterrupts()
+	defer riscv.EnableInterrupts(mask)
+
+	intr := getInterruptForHandlerID(HandlerID(i.num))
+
+	// map mapRegister to intr
+	if err := HandlerID(i.num).set(intr); err != nil {
+		return err
+	}
+
+	esp.INTERRUPT_CORE0.CPU_INT_ENABLE.SetBits(1 << intr)
+	esp.INTERRUPT_CORE0.CPU_INT_TYPE.SetBits(1 << intr)
+	// Set threshold to 5
+	reg := (*volatile.Register32)(unsafe.Pointer((uintptr(unsafe.Pointer(&esp.INTERRUPT_CORE0.CPU_INT_PRI_0)) + uintptr(intr)*4)))
+	reg.Set(5)
+
+	esp.INTERRUPT_CORE0.CPU_INT_CLEAR.SetBits(1 << intr)
+	esp.INTERRUPT_CORE0.CPU_INT_CLEAR.ClearBits(1 << intr)
+
+	riscv.Asm("fence")
+	return nil
+}
+
+// getHandlerIDsForInterrupt return registered codes for interrupt
+func getHandlerIDsForInterrupt(interrupt int) []HandlerID {
+	switch interrupt {
+	case 1:
+		return []HandlerID{GPIO}
+	case 2:
+		return []HandlerID{UART0}
+	case 3:
+		return []HandlerID{UART1}
+	default:
+		return []HandlerID{
+			USB,
+			I2C0,
+			I2S1,
+			TIMG0,
+			TIMG1,
+			UHCI0,
+			RMT,
+			SPI1,
+			SPI2,
+			TWAI,
+			RNG,
+			WIFI,
+			BT,
+			WIFI_BT_COMMON,
+			BT_BASEBAND,
+			BT_LC,
+			RSA,
+			AES,
+			SHA,
+			HMAC,
+			DS,
+			GDMA,
+			SYSTIMER,
+			SARADC,
+		}
+	}
+}
+
+// getInterruptForHandlerID return registered codes for interrupt
+func getInterruptForHandlerID(ID HandlerID) int {
+	switch ID {
+	case GPIO:
+		return 1
+	case UART0:
+		return 2
+	case UART1:
+		return 3
+	default:
+		return 4
+	}
+}
+
+func (h HandlerID) set(interrupt int) error {
+	switch h {
+	case UART0:
+		esp.INTERRUPT_CORE0.UART_INTR_MAP.Set(uint32(interrupt))
+	case UART1:
+		esp.INTERRUPT_CORE0.UART1_INTR_MAP.Set(uint32(interrupt))
+	case GPIO:
+		esp.INTERRUPT_CORE0.GPIO_INTERRUPT_PRO_MAP.Set(uint32(interrupt))
+	default:
+		return ErrInterruptNotImplemented
+	}
+	return nil
+}
+
+//export handleInterrupt
+func handleInterrupt() {
+	cause := riscv.MCAUSE.Get()
+	interruptNumber := uint32(cause & 0x1f)
+
+	if cause&(1<<31) != 0 && interruptNumber > 0 {
+		// disable this interrupt
+		interruptBit := uint32(1 << interruptNumber)
+		esp.INTERRUPT_CORE0.CPU_INT_ENABLE.ClearBits(interruptBit)
+
+		// reset this interrupt
+		if esp.INTERRUPT_CORE0.CPU_INT_TYPE.Get()&(interruptBit) != 0 {
```

Comment:
The extra parentheses are unnecessary.

```suggestion
		if esp.INTERRUPT_CORE0.CPU_INT_TYPE.Get() & interruptBit != 0 {
```